samples/DatabaseContainers/DatabaseContainers.ApiService/ApiEndpoints.cs

```diff
@@ -0,0 +1,97 @@
+﻿using Dapper;
+using Microsoft.Data.SqlClient;
+using MySqlConnector;
+using Npgsql;
+
+namespace DatabaseContainers.ApiService;
+
+public static class ApiEndpoints
+{
+    public static WebApplication MapTodosApi(this WebApplication app)
+    {
+        app.MapGet("/todos", async (NpgsqlDataSource db) =>
+        {
+            const string sql = """
+                SELECT Id, Title, IsComplete
+                FROM Todos
+                """;
+            var connection = db.CreateConnection();
+            return await connection.QueryAsync<Todo>(sql);
+        });
+
+        app.MapGet("/todos/{id}", async (int id, NpgsqlDataSource db) =>
+        {
+            const string sql = """
+                SELECT Id, Title, IsComplete
+                FROM Todos
+                WHERE Id = @id
+                """;
+            var connection = db.CreateConnection();
+            return await connection.QueryFirstOrDefaultAsync<Todo>(sql, new { id }) is { } todo
+                ? Results.Ok(todo)
+                : Results.NotFound();
+        });
+
+        return app;
+    }
+
+    public static WebApplication MapCatalogApi(this WebApplication app)
+    {
+        app.MapGet("/catalog", async (MySqlDataSource db) =>
```

Comment:
```suggestion
        app.MapGet("/catalog", async (MySqlConnection connection) =>
```

`AddMySqlDataSource` (and `AddNpgsqlDataSource`) also add `ServiceDescriptor`s that can provide a `MySqlConnection` (or `NpgsqlConnection`). So instead of having the data source injected, then calling `CreateConnection` on it, one can more simply use the transient `MySqlConnection` that's available via DI. This also has the benefit that it's automatically disposed for you.

This technique could also also be applied with `NpgsqlConnection` above in those API routes.